### PR TITLE
Adds Config file to populate application data in DB

### PR DESCRIPTION
**Pull request by SM43**
Previously, while running migration to create tables we used to insert data
into tags, categories and catalogs tables but now we keep that data in config
file at root of repo and reads from that file and add to db while starting
the api server.

Signed-off-by: Shivam Mukhade <smukhade@redhat.com>

### CODE REVIEW

File: api/.env.dev [hunk 1]
```diff
@@ -5,3 +5,5 @@ POSTGRESQL_PORT="5432"
 POSTGRESQL_DATABASE="hub"
 POSTGRESQL_USER="postgres"
 POSTGRESQL_PASSWORD="postgres"
+
+CONFIG_FILE_URL="https://raw.githubusercontent.com/sm43/hub/read-config-from-repo/config.yaml"
```
Comment (PuneetPunamiya) [inline api/.env.dev:9]: Should this yaml point to your repo ??

File: api/design/status.go [hunk 2]
```diff
@@ -1,3 +1,17 @@
+// Copyright © 2020 The Tekton Authors.
```
Comment (PuneetPunamiya) [inline api/design/status.go:1]: Should this change come under this commit ?

Comment (sthaha) [reply api/design/status.go:1]: yes, for all new files

File: api/.env.dev (hunk 1 above)
Comment (sthaha) [inline api/.env.dev:9]: I think it is better to point to the `file://../config.yaml` ?

File: api/cmd/api/main.go [hunk 3]
```diff
@@ -63,6 +67,27 @@ func main() {
...
+		// Read the config file from the repository to add into the DB
+		data, err := app.ReadConfig(api.ConfigFileURL())
+		if err != nil {
+			logger.Fatalf("Failed to read config file: %v", err)
+		}
+
+		// Viper instance to read data from file
+		v := viper.New()
+		v.SetConfigType("yaml")
+		v.ReadConfig(bytes.NewBuffer(data))
```
Comment (sthaha) [inline api/cmd/api/main.go:80]: I feel this code should move into api.Config itself

File: api/cmd/api/main.go [hunk 4]
```diff
@@ -63,6 +64,14 @@ func main() {
 		defer api.Cleanup()
 	}
 
+	{
```
Comment (sthaha) [inline api/cmd/api/main.go:67]: we don't need the `{ } ` do we?

File: api/cmd/api/main.go [hunk 5]
```diff
@@ -63,6 +64,14 @@ func main() {
...
+		// Populate Tables
+		initializer := initializer.New(api)
```
Comment (sthaha) [inline api/cmd/api/main.go:69]: should we skip this in `dev` mode :) ?

Comment (SM43) [reply api/cmd/api/main.go]: we will address this later as discussed.

File: api/config/10-db-secret.yaml [hunk 6]
```diff
@@ -21,7 +21,7 @@ metadata:
     app: db
 type: Opaque
 stringData:
-  POSTGRESQL_DATABASE: hub
-  POSTGRESQL_USER: postgres
-  POSTGRESQL_PASSWORD: postgres
-  POSTGRESQL_PORT: "5432"
+  POSTGRES_DB: hub
```
Comment (sthaha) [inline api/config/10-db-secret.yaml:24]: why is this change needed?

File: api/config/21-api-deployment.yaml [hunk 7]
```diff
@@ -53,25 +53,30 @@ spec:
             successThreshold: 1
             timeoutSeconds: 1
           env:
-            - name: POSTGRESQL_HOST
+            - name: POSTGRES_HOST
```
Comment (sthaha) [inline api/config/21-api-deployment.yaml:56]: isn't this an unrelated change?

File: api/pkg/app/app.go [hunk 8]
```diff
@@ -218,3 +240,14 @@ func initLogger(mode EnvMode) (*zap.SugaredLogger, error) {
 	}
 	return log.Sugar(), nil
 }
+
+// configFileURL will look for CONFIG_FILE_URL to be defined among
+// environment variables
+func configFileURL() (string, error) {
+
+	if val := viper.GetString("CONFIG_FILE_URL"); val != "" {
+		return val, nil
```
Comment (sthaha) [inline api/pkg/app/app.go:249]: could change this to 
```
val = viper.GetString ..
if  val == "" {
   return "", error 
}
return val, nil
```
so that the exception is inside the if

File: api/pkg/db/initializer/initiailizer.go [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+
+	for _, c := range i.data.Categories {
+		cat := &model.Category{Name: c.Name}
+		if err := i.db.Where(&cat).FirstOrCreate(&cat).Error; err != nil {
+			i.log.Error(err)
+			return err
+		}
+		for _, t := range c.Tags {
+			tag := &model.Tag{Name: t, CategoryID: cat.ID}
+			if err := db.Model(&model.Tag{}).Where(&model.Tag{Name: t}).FirstOrCreate(&tag).Error; err != nil {
```
Comment (sthaha) [inline api/pkg/db/initializer/initiailizer.go:59]: how about breaking this long line into 

```
err := db.Model....;
if err != nil {
}
```

File: api/pkg/db/initializer/initiailizer.go [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+	for _, c := range i.data.Catalogs {
+		cat := &model.Catalog{
+			Name:       c.Name,
+			Type:       c.Type,
+			URL:        c.URL,
+			Owner:      c.Owner,
+			Revision:   c.Revision,
+			ContextDir: c.ContextDir,
+		}
+		if err := db.Where(&model.Catalog{Name: c.Name, Type: c.Type}).FirstOrCreate(&cat).Error; err != nil {
```
Comment (sthaha) [inline api/pkg/db/initializer/initiailizer.go:86]: same, could we break the long line?

File: api/pkg/db/model/model.go [hunk 11]
```diff
@@ -29,7 +29,7 @@ type (
 		gorm.Model
 		Name       string `gorm:"size:100;not null;unique"`
 		Category   Category
-		CategoryID int
+		CategoryID uint
```
Comment (sthaha) [inline api/pkg/db/model/model.go:32]: does this need any migration changes? I guess not but could you please confirm?

File: api/test/config/env.test [hunk 12]
```diff
@@ -1,6 +1,9 @@
 ENVIRONMENT="test"
-POSTGRESQL_HOST="localhost"
-POSTGRESQL_PORT="5432"
-POSTGRESQL_DATABASE="hub_test"
-POSTGRESQL_USER="postgres"
-POSTGRESQL_PASSWORD="postgres"
+
```
Comment (sthaha) [inline api/test/config/env.test:2]: unrelated change ?

File: api/test/config/env.test [hunk 13]
```diff
@@ -1,6 +1,9 @@
...
+POSTGRES_HOST="localhost"
+POSTGRES_PORT="5432"
+POSTGRES_DB="hub_test"
+POSTGRES_USER="postgres"
+POSTGRES_PASSWORD="postgres"
+
+CONFIG_FILE_URL="file://../../../../config.yaml"
```
Comment (sthaha) [inline api/test/config/env.test:9]: its better to keep another config.yaml file for test, that can be empty

File: test/presubmit-tests.sh [hunk 14]
```diff
@@ -57,14 +57,14 @@ set-pg-passwd() {
 api-unittest(){
   install-postgres
   source $API_DIR/test/config/env.test
-  set-pg-passwd "$POSTGRESQL_PASSWORD"
+  set-pg-passwd "$POSTGRES_PASSWORD"
```
Comment (sthaha) [inline test/presubmit-tests.sh:60]: unrelated change.

File: api/pkg/db/model/model.go (hunk 11 above)
Comment (SM43) [reply api/pkg/db/model/model.go:32]: Ideally yes it needs a migration but it won't be breaking anything because `CategoryID` refers to Tag `Id` which is the primary key and autogenerated and won't be a signed int. And we don't access category Id except for preloading.

Comment (SM43) [reply api/pkg/db/model/model.go:32]: but yes will write a migration anyway. Thanks for reminding :)

Comment (SM43) [reply api/pkg/db/model/model.go:32]: I checked this one, queries are the same for `uint` and `int` as Postgres have the `integer` data type. so, skipping the migration.

File: api/test/config/env.test [hunk 15]
```diff
@@ -1,6 +1,9 @@
...
-POSTGRESQL_USER="postgres"
-POSTGRESQL_PASSWORD="postgres"
+
+POSTGRES_HOST="localhost"
+POSTGRES_PORT="5432"
+POSTGRES_DB="hub_test"
+POSTGRES_USER="postgres"
+POSTGRES_PASSWORD="postgres"
+
+CONFIG_FILE_URL="file://../../../test/config/config.yaml"
```
Comment (sthaha) [inline api/test/config/env.test:9]: @SM43 could you please use a different file config.yaml file for test?

File: api/config/20-api-configmap.yaml [hunk 16]
```diff
@@ -0,0 +1,23 @@
...
+
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: api
+  namespace: tekton-hub
+  labels:
+    app: api
+data:
+  CONFIG_FILE_URL: https://raw.githubusercontent.com/tektoncd/hub/master/config.yaml
```
Comment (sthaha) [inline api/config/20-api-configmap.yaml:23]: is this a valid yaml?

Comment (SM43) [reply api/config/20-api-configmap.yaml:23]: Once the pr is mergered then it will be

File: api/pkg/db/initializer/initiailizer.go [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+	return nil
+}
+
+func (i *Initializer) addCategories() error {
+
+	db := i.db
+
+	// Checks if tables exists
+	if !db.HasTable(&model.Category{}) || !db.HasTable(model.Tag{}) {
+		return fmt.Errorf("Categories or Tags Table not found")
```
Comment (sthaha) [inline api/pkg/db/initializer/initiailizer.go:48]: could you please use lowercase of `errors` 🙏
